File: src/app/static/src/stats.ts
// Client-side aggregations — all stats happen here since server can't see ciphertext.

import type { ActivityLevel, FoodEntryPayload, Gender, MeasurementPayload, ProfilePayload, WaterEntryPayload } from "./types";

export function today(): string {
  return new Date().toISOString().slice(0, 10);
}

export function isoDaysAgo(days: number): string {
  const d = new Date();
  d.setDate(d.getDate() - days);
  return d.toISOString().slice(0, 10);
}

export interface MacroTotals {
  calories: number;
  protein_g: number;
  fat_g: number;
  carbs_g: number;
}

export function sumFood(entries: FoodEntryPayload[]): MacroTotals {
  return entries.reduce<MacroTotals>(
    (acc, e) => ({
      calories: acc.calories + (e.calories || 0),
      protein_g: acc.protein_g + (e.protein_g || 0),
      fat_g: acc.fat_g + (e.fat_g || 0),
      carbs_g: acc.carbs_g + (e.carbs_g || 0),
    }),
    { calories: 0, protein_g: 0, fat_g: 0, carbs_g: 0 },
  );
}

export function sumWater(entries: WaterEntryPayload[]): number {
  return entries.reduce((a, e) => a + (e.amount_ml || 0), 0);
}

// TDEE = BMR * activity multiplier (Mifflin-St Jeor).
export function calcTDEE(p: ProfilePayload): number | null {
  if (!p.gender || !p.weight_kg || !p.height_cm || !p.birth_date || !p.activity_level) return null;
  const age = Math.floor((Date.now() - new Date(p.birth_date).getTime()) / (365.25 * 24 * 3600 * 1000));
  const bmr =
    p.gender === "male"
      ? 10 * p.weight_kg + 6.25 * p.height_cm - 5 * age + 5
      : 10 * p.weight_kg + 6.25 * p.height_cm - 5 * age - 161;
  const mult: Record<ActivityLevel, number> = {
    sedentary: 1.2, light: 1.375, moderate: 1.55, active: 1.725, very_active: 1.9,
  };
  return Math.round(bmr * mult[p.activity_level]);
}

export function latestWeight(measurements: MeasurementPayload[]): number | null {
  const withWeight = measurements.filter((m) => m.weight_kg !== undefined && m.weight_kg !== null);
  withWeight.sort((a, b) => new Date(b.measured_at).getTime() - new Date(a.measured_at).getTime());
  return withWeight[0]?.weight_kg ?? null;
}

export function genderLabel(g: Gender | undefined): string {
  if (g === "male") return "Male";
  if (g === "female") return "Female";
  return "—";
}

export function isProfileComplete(p: ProfilePayload | undefined): boolean {
  if (!p) return false;
  return !!(p.gender && p.weight_kg && p.height_cm && p.birth_date && p.activity_level);
}

export function activityLabel(a: ActivityLevel | undefined): string {
  switch (a) {
    case "sedentary": return "Sedentary";
    case "light": return "Light";
    case "moderate": return "Moderate";
    case "active": return "Active";
    case "very_active": return "Very active";
    default: return "—";
  }
}

File: android/app/src/main/kotlin/com/trackhub/util/Stats.kt
package com.trackhub.util

import com.trackhub.api.FoodEntryPayload
import com.trackhub.api.MeasurementPayload
import com.trackhub.api.ProfilePayload
import com.trackhub.api.WaterEntryPayload
import java.time.LocalDate
import java.time.Period
import java.time.format.DateTimeFormatter

data class MacroTotals(
    val calories: Double,
    val proteinG: Double,
    val fatG: Double,
    val carbsG: Double,
)

object Stats {
    fun sumFood(entries: List<FoodEntryPayload>): MacroTotals {
        var c = 0.0; var p = 0.0; var f = 0.0; var cb = 0.0
        for (e in entries) { c += e.calories; p += e.proteinG; f += e.fatG; cb += e.carbsG }
        return MacroTotals(c, p, f, cb)
    }

    fun sumWater(entries: List<WaterEntryPayload>): Int = entries.sumOf { it.amountMl }

    fun latestWeight(measurements: List<MeasurementPayload>): Double? {
        return measurements
            .filter { it.weightKg != null }
            .sortedByDescending { it.measuredAt }
            .firstOrNull()?.weightKg
    }

    /** Mifflin-St Jeor. Returns null when any required field is missing. */
    fun calcTdee(p: ProfilePayload): Int? {
        val g = p.gender ?: return null
        val w = p.weightKg ?: return null
        val h = p.heightCm ?: return null
        val birth = p.birthDate ?: return null
        val activity = p.activityLevel ?: return null
        val age = try {
            val bd = LocalDate.parse(birth, DateTimeFormatter.ISO_LOCAL_DATE)
            Period.between(bd, LocalDate.now()).years
        } catch (_: Exception) { return null }
        val bmr = if (g == "male") 10 * w + 6.25 * h - 5 * age + 5
                  else 10 * w + 6.25 * h - 5 * age - 161
        val mul = when (activity) {
            "sedentary" -> 1.2; "light" -> 1.375; "moderate" -> 1.55
            "active" -> 1.725; "very_active" -> 1.9
            else -> return null
        }
        return (bmr * mul).toInt()
    }

    fun activityLabel(a: String?): String = when (a) {
        "sedentary" -> "Сидячий"
        "light" -> "Лёгкая"
        "moderate" -> "Умеренная"
        "active" -> "Активный"
        "very_active" -> "Очень активный"
        else -> "—"
    }

    fun genderLabel(g: String?): String = when (g) {
        "male" -> "Мужской"; "female" -> "Женский"; else -> "—"
    }
}
